src/shape.rs

```diff
@@ -660,237 +677,214 @@ impl ShapeLine {
-                                    word_range_width = glyph_size;
+                                    fit_x = line_width - glyph_width;
+                                    word_range_width = glyph_width;
                                     fitting_start = (i, glyph_i + 1);
                                 }
                             }
                         } else {
                             // Wrap::Word
-                            let mut prev_word_width = None;
-                            if word.blank && number_of_blanks > 0 {
-                                // current word causing a wrap is a space so we ignore it
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                            let mut trialing_space_width = None;
+                            if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        trialing_space_width =
+                                            Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
                                 }
                             }
-                            if let Some(width) = prev_word_width {
-                                word_ranges.push((
-                                    (i, 0),
+                            if let Some(width) = trialing_space_width {
+                                self.add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
+                                    (i + 2, 0),
                                     fitting_start,
                                     word_range_width - width,
                                     number_of_blanks,
-                                ));
+                                );
                             } else {
-                                word_ranges.push((
+                                self.add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
                                     (i + 1, 0),
```

Comment:
Same as above, why is it `i + 1` here, but just `i` in the other direction?

Reply:
Here is how the algorithm currently works:

- When we are shaping a some text. Depending on the directionality of the first character we decide the direction of the whole paragraph. So if the paragraph starts with an RTL character and the rest is all LTR, we decide that that whole paragraph is RTL.

- We then create `spans` of text. Each span is created based on the text direction. So, if the whole paragraph is LTR, the whole paragraph is only one span. But if you have BiDi text, which means a mix of LTR and RTL, then the paragraph is chopped up in smaller spans. **Each span always has only one direction. Either LTR or RTL.**

- When we are shaping an LTR span inside of an RTL paragraph (or the other way around) we store the words for that span in a reverse order. It happens [here](https://github.com/pop-os/cosmic-text/blob/c4a8d521f60ad35f95b083256ae993de2537d3a0/src/shape.rs#L419).

- Now, in the `ShapeLine::layout()` function, we iterate over these spans and find out what fits inside the line_width. If the direction of the paragraph is the same as the direction of the span, this process is pretty straight forward. But for the cases that the directions differ we have to iterate through them using `.rev()`. 

- In the layout function, we find the ranges of words inside of a span that fits inside a line. So, if we have a long span of words. It will take multiple lines. So, the way we are saving it is `(span_index, start, end)`. Each `VisualLine` has a `Vec<(span_index, start, end)>`. 

- Later we create `LayoutLine` using the ranges of words inside the `VisualLine`. `LayoutLine` contains the information about the glyphs and their positions.

As an example Let's say we have a paragraph that starts with a very long English text (LTR) and has some Farsi in it too (RTL). Let's say our desired output should look like this:
```
LLL LL LLLL LLLL LL LLL LL
LLLL LLLLL LLLL R RRR
RR RRRR LLLL LLL.
```
Now we have 3 spans in this paragraph:
```
span 0: LLL LL LLLL LLLL LL LLL LL LLLL LLLLL LLLL 
span 1: RR RRRR  R RRR 
span 2: LLLL LLL.
```
The first visual line would be pretty easy: 
```
Visual Line 0: [(0, 0, 7)]
```
It is saying span 0, from the first word till the 7th word (index 0 to 7, not inclusive on 7).

The second visual line is different! It starts with the rest of the first span, but when it gets to the RTL text the indicies get a little trickier.

If we didn't have a line break the RTL text would look like this: `RR RRRR  R RRR`, but now that we have a line break the two right-most words would fit but the two left-most words would be placed in the next line.

So the visual line would look like:
```
Visual Line 1: [(0,7,10), (1, 2, 4)]
```
and the last visual line would be:
```
Visual Line 2: [(1, 0, 2), (2, 0, 2)]
```



Now, when we are iterating over words in the case that the paragraph direction matches the span direction, the `i` would mean the _next_ word (visually). But in case of incongruent directions, the `i` means _previous_ word (visually).
Since we also are going in reverse direction, if we find a word that doesn't fit we store `(span_index, i+1, fitting_start)` as the range of words that would fit. Since we are moving backwards `fitting_start` is bigger than `i`.

Now if the previous word was a blank, then we ignore that and save `i+2` as the start of that word range. 

**Notice that the start of a word range is inclusive.**

If the direction of the span and the paragraph are the same and if we find a word that doesn't fit we store `(span_index, fitting_start, i)`. If the previous word was a blank we store `(span_index, fitting_start, i-1)`. Since in this case we are iterating in the correct direction (and not reversed) then ` i > fitting_start`.

**Notice that the end of the word range is not inclusive.**

Ok. This is the basic algorithm. However, since we also want to be able to wrap in a character level, we need to also say from which glyph to which glyph it would fit inside a line.

So our Visual Lines are actually `(span_index, (starting_word_index, starting_word_glyph_index), (ending_word_index, ending_word_glyph_index)`. Also note that it is possible to have a long word that is broken over multiple lines.

So if you have a word range that says `(0, (0,0), (0,10))`. That means the algorithm could only fit 10 glyphs from the first word of the first span.

If the word range says `(0, (0,0), (4,0))`, it means we were able to fit the first 4 words of the first span.

If the word range says `(0, (2,1), (3,5))`, it means from the second glyph of the third word, until the 5th glyph of the forth word.

I hope this helped.